File: verifiers/typescript/conformance-runner.ts
// Cross-language conformance runner for the standalone TypeScript VCC L1
// verifier (sdk/typescript).
//
// Loads the SAME committed corpus the in-repo conformance suite uses
// (`vectors/`), runs THIS independent verifier over every positive
// and negative vector, and asserts the result matches the pinned outcome — the
// outcome the reference verifier is pinned to in
// `src/lib/vcc/conformance.test.ts`. It is the direct TypeScript analog of
// `sdk/python/conformance_runner.py`.
//
// This is the machine-checkable half of the §50 "Interoperable" gate: an
// independent, site-decoupled verifier producing the same result on the same
// vectors. OFFLINE — only local files, no network.
//
// Exit 0 iff every vector matches. Prints a per-vector table and a summary.

import { existsSync, readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { canonicalize, canonicalBytes } from "./src/canonicalize.js";
import { VCC_CALC_URN_PREFIX } from "./src/constants.js";
import { verifyVccEnvelope } from "./src/verify.js";
import type {
  VccCertificateStatus,
  VccIssuerKey,
  VccIssuerKeySet,
  VccStatement,
} from "./src/types.js";

const HERE = dirname(fileURLToPath(import.meta.url));

/**
 * Locate the committed corpus at `vectors` by walking up from this
 * file. This works whether the runner executes from source (sdk/typescript/) or
 * from the compiled `dist/`, so the "path relative to ../../vectors"
 * contract holds without hard-coding a level count.
 */
function findVectorDir(start: string): string {
  let dir = start;
  for (let i = 0; i < 8; i++) {
    const candidate = join(dir, "vectors");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error(
    "could not locate vectors above " + start +
      " — run this from a checkout of the protocol repo",
  );
}

const VECTOR_DIR = findVectorDir(HERE);
const NEG_DIR = join(VECTOR_DIR, "negative");
const KEYSET_FILE = join(VECTOR_DIR, "test-key.json");

function loadJson(p: string): any {
  return JSON.parse(readFileSync(p, "utf8"));
}

function baseKeyset(): VccIssuerKeySet {
  return loadJson(KEYSET_FILE).keyset as VccIssuerKeySet;
}

/** Mirror keysetFor() in conformance.test.ts / keyset_for() in the Python runner. */
function keysetFor(name: string | null | undefined): VccIssuerKeySet {
  const ks = baseKeyset();
  if (name === "revoked") {
    return {
      issuer: ks.issuer,
      keys: ks.keys.map((k: VccIssuerKey) => ({
        ...k,
        status: "revoked" as VccIssuerKey["status"],
      })),
    };
  }
  if (name === "wrong-algorithm") {
    return {
      issuer: ks.issuer,
      keys: ks.keys.map((k: VccIssuerKey) => ({
        ...k,
        algorithm: "rsa" as VccIssuerKey["algorithm"],
      })),
    };
  }
  return ks;
}

async function sha256Hex(bytes: Uint8Array): Promise<string> {
  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", bytes as BufferSource));
  let hex = "";
  for (const b of digest) hex += b.toString(16).padStart(2, "0");
  return hex;
}

async function computeStatementIdHex(statement: VccStatement): Promise<string> {
  const clone = structuredClone(statement) as { subject: { id?: string } };
  delete clone.subject.id;
  return sha256Hex(canonicalBytes(clone));
}

type Result = [ok: boolean, detail: string];

// ── Positive corpus ───────────────────────────────────────────────────────────

async function checkPositive(file: string): Promise<Result> {
  const v = loadJson(file);
  const exp = v.expected?.l1 ?? {};
  const res = await verifyVccEnvelope(v.envelope, baseKeyset());

  const problems: string[] = [];
  if (exp.cryptographicValidity != null && res.cryptographicValidity !== exp.cryptographicValidity) {
    problems.push(
      `cryptographicValidity=${res.cryptographicValidity} want ${exp.cryptographicValidity}`,
    );
  }
  if (
    exp.trustedAtVerificationTime != null &&
    res.trustedAtVerificationTime !== exp.trustedAtVerificationTime
  ) {
    problems.push(
      `trustedAtVerificationTime=${res.trustedAtVerificationTime} want ${exp.trustedAtVerificationTime}`,
    );
  }
  if (res.errors.length > 0) {
    problems.push(`errors not empty: ${JSON.stringify(res.errors)}`);
  }
  if (!res.cryptographicValidity && exp.cryptographicValidity !== false) {
    const failed = Object.entries(res.checks)
      .filter(([, val]) => !val)
      .map(([k]) => k);
    problems.push(`unexpected failed checks: ${JSON.stringify(failed)}`);
  }
  return [problems.length === 0, problems.join("; ")];
}

// ── Negative corpus ────────────────────────────────────────────────────────────

async function checkNegative(file: string): Promise<Result> {
  const v = loadJson(file);
  const exp = v.expected;
  const ks = keysetFor(exp.keyset);
  const certStatus = (exp.certificateStatus as VccCertificateStatus | undefined) ?? null;
  const res = await verifyVccEnvelope(v.envelope, ks, { certificateStatus: certStatus });
  const checks = res.checks as unknown as Record<string, boolean>;

  const problems: string[] = [];

  // Every pinned failed check must be false in our result.
  for (const check of exp.l1FailedChecks ?? []) {
    if (checks[check] !== false) {
      problems.push(`check ${check}=${checks[check]} want false`);
    }
  }

  if (exp.l1CryptographicValidity != null) {
    if (res.cryptographicValidity !== exp.l1CryptographicValidity) {
      problems.push(
        `cryptographicValidity=${res.cryptographicValidity} want ${exp.l1CryptographicValidity}`,
      );
    }
    if (exp.l1CryptographicValidity === false && res.errors.length === 0) {
      problems.push("expected non-empty errors");
    }
  }

  if (exp.l1TrustedAtVerificationTime != null) {
    if (res.trustedAtVerificationTime !== exp.l1TrustedAtVerificationTime) {
      problems.push(
        `trustedAtVerificationTime=${res.trustedAtVerificationTime} want ${exp.l1TrustedAtVerificationTime}`,
      );
    }
  }

  return [problems.length === 0, problems.join("; ")];
}

// ── JCS byte-parity: the load-bearing interoperability check ──────────────────
// Each positive vector carries the reference-produced canonical string
// (`canonicalStatement`) and its sha256. This verifier's JCS must reproduce BOTH
// exactly, and the content-addressed statementId must match subject.id.

async function checkJcsParity(file: string): Promise<Result> {
  const v = loadJson(file);
  const st = v.statement as VccStatement;
  const problems: string[] = [];
  if ("canonicalStatement" in v && canonicalize(st) !== v.canonicalStatement) {
    problems.push("JCS bytes differ from reference canonicalStatement");
  }
  if ("canonicalStatementSha256" in v) {
    const sha = await sha256Hex(canonicalBytes(st));
    if (sha !== v.canonicalStatementSha256) {
      problems.push("canonical sha256 differs from pinned");
    }
  }
  const idHex = await computeStatementIdHex(st);
  if (VCC_CALC_URN_PREFIX + idHex !== st.subject.id) {
    problems.push("statementId differs from subject.id");
  }
  return [problems.length === 0, problems.join("; ")];
}

async function main(): Promise<number> {
  const positives = readdirSync(VECTOR_DIR)
    .filter((f) => f.endsWith(".json") && f !== "test-key.json")
    .sort()
    .map((f) => join(VECTOR_DIR, f));
  const negatives = readdirSync(NEG_DIR)
    .filter((f) => f.endsWith(".json") && f !== "index.json")
    .sort()
    .map((f) => join(NEG_DIR, f));

  let passed = 0;
  let total = 0;
  const rows: Array<[kind: string, name: string, ok: boolean, detail: string]> = [];

  let jcsAllOk = true;
  for (const f of positives) {
    const [ok, detail] = await checkJcsParity(f);
    jcsAllOk = jcsAllOk && ok;
    total++;
    if (ok) passed++;
    rows.push(["JCS", basename(f), ok, detail]);
  }
  for (const f of positives) {
    const [ok, detail] = await checkPositive(f);
    total++;
    if (ok) passed++;
    rows.push(["POS", basename(f), ok, detail]);
  }
  for (const f of negatives) {
    const [ok, detail] = await checkNegative(f);
    total++;
    if (ok) passed++;
    rows.push(["NEG", basename(f), ok, detail]);
  }

  const width = Math.max(...rows.map((r) => r[1].length));
  console.log("VCC cross-language conformance — TypeScript SDK verifier vs pinned reference outcome");
  console.log(`corpus: ${VECTOR_DIR}`);
  console.log("-".repeat(width + 24));
  for (const [kind, name, ok, detail] of rows) {
    const mark = ok ? "PASS" : "FAIL";
    let line = `[${kind}] ${name.padEnd(width)}  ${mark}`;
    if (!ok) line += `  <- ${detail}`;
    console.log(line);
  }
  console.log("-".repeat(width + 24));
  console.log(`MATCH: ${passed}/${total} checks match the pinned (reference) outcome`);
  console.log(`JCS byte-for-byte parity with reference on all positives: ${jcsAllOk}`);
  return passed === total ? 0 : 1;
}

function basename(p: string): string {
  return p.split("/").pop() ?? p;
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err);
    process.exit(2);
  });

File: verifiers/python/conformance_runner.py
"""Cross-language conformance runner for the Python VCC L1 verifier.

Loads the SAME committed corpus the TypeScript conformance suite uses
(`src/lib/vcc/vectors/`), runs the independent Python verifier over every
positive and negative vector, and asserts the Python result matches the pinned
outcome — the outcome the TS reference verifier is pinned to in
`src/lib/vcc/conformance.test.ts`.

This is the machine-checkable half of the §50 "Interoperable" gate: two
independent verifiers, same result on the same vectors. OFFLINE — only local
files, no network.

Exit 0 iff every vector matches. Prints a per-vector table and a summary.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import hashlib  # noqa: E402

sys.path.insert(0, str(Path(__file__).resolve().parent))
from vcc_verifier import (  # noqa: E402
    canonical_bytes,
    canonicalize,
    compute_statement_id_hex,
    verify_vcc_envelope,
)

REPO_ROOT = Path(__file__).resolve().parents[2]
VECTOR_DIR = REPO_ROOT / "vectors"
NEG_DIR = VECTOR_DIR / "negative"
KEYSET_FILE = VECTOR_DIR / "test-key.json"


def load_json(p: Path):
    return json.loads(p.read_text(encoding="utf-8"))


def base_keyset() -> dict:
    return load_json(KEYSET_FILE)["keyset"]


def keyset_for(name: str | None) -> dict:
    """Mirror keysetFor() in conformance.test.ts."""
    ks = base_keyset()
    if name == "revoked":
        keys = [dict(k, status="revoked") for k in ks["keys"]]
        return {"issuer": ks["issuer"], "keys": keys}
    if name == "wrong-algorithm":
        keys = [dict(k, algorithm="rsa") for k in ks["keys"]]
        return {"issuer": ks["issuer"], "keys": keys}
    if name == "issuer-mismatch":
        return {"issuer": "https://attacker.example", "keys": ks["keys"]}
    if name == "future-key":
        keys = [dict(k, validFrom="2099-01-01T00:00:00Z") for k in ks["keys"]]
        return {"issuer": ks["issuer"], "keys": keys}
    if name == "expired-key":
        keys = [dict(k, validUntil="2020-01-01T00:00:00Z") for k in ks["keys"]]
        return {"issuer": ks["issuer"], "keys": keys}
    return ks


# ── Positive corpus ───────────────────────────────────────────────────────────


def check_positive(file: Path) -> tuple[bool, str]:
    v = load_json(file)
    exp = v.get("expected", {}).get("l1", {})
    res = verify_vcc_envelope(v["envelope"], base_keyset())

    problems = []
    want_crypto = exp.get("cryptographicValidity")
    if want_crypto is not None and res.cryptographicValidity != want_crypto:
        problems.append(f"cryptographicValidity={res.cryptographicValidity} want {want_crypto}")
    want_trust = exp.get("trustedAtVerificationTime")
    if want_trust is not None and res.trustedAtVerificationTime != want_trust:
        problems.append(f"trustedAtVerificationTime={res.trustedAtVerificationTime} want {want_trust}")
    if res.errors:
        problems.append(f"errors not empty: {res.errors}")
    if not res.cryptographicValidity and want_crypto is not False:
        failed = [k for k, val in res.checks.items() if not val]
        problems.append(f"unexpected failed checks: {failed}")
    return (not problems, "; ".join(problems))


# ── Negative corpus ────────────────────────────────────────────────────────────


def check_negative(file: Path) -> tuple[bool, str]:
    v = load_json(file)
    exp = v["expected"]
    ks = keyset_for(exp.get("keyset"))
    cert_status = exp.get("certificateStatus")
    res = verify_vcc_envelope(v["envelope"], ks, certificate_status=cert_status)

    problems = []

    # Every pinned failed check must be false in the Python result.
    for check in exp.get("l1FailedChecks") or []:
        if res.checks.get(check) is not False:
            problems.append(f"check {check}={res.checks.get(check)} want False")

    want_crypto = exp.get("l1CryptographicValidity")
    if want_crypto is not None and res.cryptographicValidity != want_crypto:
        problems.append(f"cryptographicValidity={res.cryptographicValidity} want {want_crypto}")
        # TS also asserts errors non-empty when cryptographicValidity=false
    if want_crypto is False and not res.errors:
        problems.append("expected non-empty errors")

    want_trust = exp.get("l1TrustedAtVerificationTime")
    if want_trust is not None and res.trustedAtVerificationTime != want_trust:
        problems.append(f"trustedAtVerificationTime={res.trustedAtVerificationTime} want {want_trust}")

    want_bound = exp.get("l1IssuerIdentityBound")
    if want_bound is not None and res.issuerIdentityBound != want_bound:
        problems.append(f"issuerIdentityBound={res.issuerIdentityBound} want {want_bound}")

    want_key_valid = exp.get("l1KeyValidAtIssuance")
    if want_key_valid is not None and res.keyValidAtIssuance != want_key_valid:
        problems.append(f"keyValidAtIssuance={res.keyValidAtIssuance} want {want_key_valid}")

    return (not problems, "; ".join(problems))


# ── JCS byte-parity: the load-bearing interoperability check ──────────────────
# Each positive vector carries the TS-produced canonical string
# (`canonicalStatement`) and its sha256. Python JCS must reproduce BOTH exactly,
# and the content-addressed statementId must match subject.id. If this diverges
# on any vector, the corpus would still "verify" but the two implementations
# would NOT be interoperable — so we surface it as its own gate.


def check_jcs_parity(file: Path) -> tuple[bool, str]:
    v = load_json(file)
    st = v["statement"]
    problems = []
    if "canonicalStatement" in v and canonicalize(st) != v["canonicalStatement"]:
        problems.append("JCS bytes differ from TS canonicalStatement")
    if "canonicalStatementSha256" in v:
        py_sha = hashlib.sha256(canonical_bytes(st)).hexdigest()
        if py_sha != v["canonicalStatementSha256"]:
            problems.append("canonical sha256 differs from pinned")
    id_hex = compute_statement_id_hex(st)
    if ("urn:vcc:calculation:sha256:" + id_hex) != st["subject"]["id"]:
        problems.append("statementId differs from subject.id")
    return (not problems, "; ".join(problems))


def main() -> int:
    positives = sorted(
        f for f in VECTOR_DIR.glob("*.json") if f.name != "test-key.json"
    )
    negatives = sorted(f for f in NEG_DIR.glob("*.json") if f.name != "index.json")

    passed = 0
    total = 0
    rows = []

    jcs_all_ok = True
    for f in positives:
        ok, detail = check_jcs_parity(f)
        jcs_all_ok = jcs_all_ok and ok
        total += 1
        passed += ok
        rows.append(("JCS", f.name, ok, detail))

    for f in positives:
        ok, detail = check_positive(f)
        total += 1
        passed += ok
        rows.append(("POS", f.name, ok, detail))

    for f in negatives:
        ok, detail = check_negative(f)
        total += 1
        passed += ok
        rows.append(("NEG", f.name, ok, detail))

    width = max(len(r[1]) for r in rows)
    print(f"VCC cross-language conformance — Python verifier vs pinned TS outcome")
    print(f"corpus: {VECTOR_DIR}")
    print("-" * (width + 24))
    for kind, name, ok, detail in rows:
        mark = "PASS" if ok else "FAIL"
        line = f"[{kind}] {name.ljust(width)}  {mark}"
        if not ok:
            line += f"  <- {detail}"
        print(line)
    print("-" * (width + 24))
    print(f"MATCH: {passed}/{total} checks match the pinned (TS reference) outcome")
    print(f"JCS byte-for-byte parity with TS on all positives: {jcs_all_ok}")
    return 0 if passed == total else 1


if __name__ == "__main__":
    raise SystemExit(main())
